### PR TITLE
[CMake] Refactor to allow an overlay only build

### PR DESCRIPTION
**Pull request by zisko**
This change makes these stdlib extras get treated like overlays.
they now only get built if SWIFT_BUILD_{STATIC,DYNAMIC}_SDK_OVERLAY
flags are added to build-script-impl, with the goal of being able to build 
"just the overlay".

It also removes the resource-dir flag. swiftc is smart enough to find
resources relative to it, and no matter where you're building (just built)
|| (external toolchain), this was just setting the resource-dir to something
swiftc grokks already.

I'm concerned this stops compiler + stdlib + StdlibCollectionUnittest builds
from being possible without building an overlay. if this is the case, we can just add 
"OR SWIFT_BUILD_DYNAMIC_STDLIB" and "OR SWIFT_BUILD_STATIC_STDLIB" in order to add STATIC or DYNAMIC to this list, but we should change the name to SWIFT_STDLIB_OR_SDK_OVERLAY_LIBRARY_BUILD_TYPES 💯.

### CODE REVIEW

**Comment by zisko** [comment]
@swift-ci please test

**Comment by zisko** [comment]
@modocache @gribozavr @erg 

Please review/ scream at me :)

File: stdlib/private/StdlibCollectionUnittest/CMakeLists.txt [hunk 1]
```diff
@@ -1,11 +1,19 @@
+set(SWIFT_SDK_OVERLAY_LIBRARY_BUILD_TYPES)
+if(SWIFT_BUILD_DYNAMIC_SDK_OVERLAY)
+  list(APPEND SWIFT_SDK_OVERLAY_LIBRARY_BUILD_TYPES SHARED)
+endif()
+if(SWIFT_BUILD_STATIC_SDK_OVERLAY)
+  list(APPEND SWIFT_SDK_OVERLAY_LIBRARY_BUILD_TYPES STATIC)
+endif()
```
Comment (modocache) [inline stdlib/private/StdlibCollectionUnittest/CMakeLists.txt]: Rather than write these seven lines of code in each of the CMakeLists.txt files in `stdlib/private`, I think you could write it once in `stdlib/private/CMakeLists.txt`, then use the variables in each of the files. That'll significantly reduce the size of this pull request!

File: cmake/modules/AddSwift.cmake [hunk 2]
```diff
@@ -223,7 +223,6 @@ function(_add_variant_swift_compile_flags
   list(APPEND result
       "-sdk" "${SWIFT_SDK_${sdk}_PATH}"
       "-target" "${SWIFT_SDK_${sdk}_ARCH_${arch}_TRIPLE}"
-      "-resource-dir" "${SWIFTLIB_DIR}")
```
Comment (modocache) [inline cmake/modules/AddSwift.cmake]: Is this change necessary to build just the overlay? Or could it be split out into a separate pull request?

Also, /cc @compnerd, in case he has any thoughts on removing this.

**Review by modocache — CHANGES_REQUESTED** [review]
Awesome! But the amount of duplication bothers me... Could you centralize the `if()` statement into a single location?

File: cmake/modules/AddSwift.cmake (hunk 2 above)
Comment (zisko) [inline cmake/modules/AddSwift.cmake]: Thanks! I'll re-do this now. It was a hack as I encountered CMake failures.

File: stdlib/private/StdlibCollectionUnittest/CMakeLists.txt (hunk 1 above)
Comment (zisko) [inline stdlib/private/StdlibCollectionUnittest/CMakeLists.txt]: Thanks! I'll re-do this now. It was a hack as I encountered CMake failures.

File: cmake/modules/AddSwift.cmake (hunk 2 above)
Comment (jrose-apple) [inline cmake/modules/AddSwift.cmake]: It is actually necessary (because the compiler probably isn't in the build directory), but it could be split out separately and committed *first* if it really mattered. I don't think it really matters, though.

**Comment by zisko** [comment]
@modocache I made some changes- seems more sane now. Thanks for looking at this!

**Comment by zisko** [comment]
@swift-ci please test

**Comment by zisko** [comment]
@swift-ci please test

**Comment by zisko** [comment]
@swift-ci please test

**Comment by erg** [comment]
Can you squash these six commits into one?

I'm sure you know how, but here's the command: ``git rebase -i HEAD~6``. Change ``pick`` to ``s`` at the beginning of lines 2-6 and write one commit message, then save it.

Otherwise looks good!

**Comment by zisko** [comment]
@swift-ci please smoke test

File: stdlib/CMakeLists.txt [hunk 3]
```diff
@@ -18,6 +18,18 @@ if(SWIFT_BUILD_STATIC_STDLIB)
   list(APPEND SWIFT_STDLIB_LIBRARY_BUILD_TYPES STATIC)
 endif()
 
+# Also make a list of "stdlib extras".
+# This will determine if StdlibUnitTest, Platform/*,
+# SwiftPrivate, etc will build. 
+# This allows this logic to work with overlay only builds.
```
Comment (gottesmm) [inline stdlib/CMakeLists.txt]: Can you improve this comment?

Suggestion:

```
# Make a list of build types for "standard library extras"
#
# A "standard library extra" is a swift library that is not
# StdlibCore or an overlay. Some examples of standard library
# extras are StdlibUnittest, Platform/*, and SwiftPrivate.
#
# The reason why this separate list is important is that it allows
# for the standard library extras to not be built when performing
# an overlay only build.
```

File: stdlib/CMakeLists.txt [hunk 4]
```diff
@@ -18,6 +18,18 @@ if(SWIFT_BUILD_STATIC_STDLIB)
...
+set(SWIFT_STDLIB_EXTRAS_BUILD_TYPES)
+if(SWIFT_BUILD_DYNAMIC_SDK_OVERLAY OR SWIFT_BUILD_DYNAMIC_STDLIB)
```
Comment (jrose-apple) [inline stdlib/CMakeLists.txt]: This isn't quite going to work; not everything in the "stdlib extras" can build without the overlays.

File: cmake/modules/AddSwift.cmake [hunk 5]
```diff
@@ -222,8 +222,7 @@ function(_add_variant_swift_compile_flags
 
   list(APPEND result
       "-sdk" "${SWIFT_SDK_${sdk}_PATH}"
-      "-target" "${SWIFT_SDK_${sdk}_ARCH_${arch}_TRIPLE}"
-      "-resource-dir" "${SWIFTLIB_DIR}")
+      "-target" "${SWIFT_SDK_${sdk}_ARCH_${arch}_TRIPLE}")
```
Comment (gottesmm) [inline cmake/modules/AddSwift.cmake]: Shouldn't this be in a different commit?

**Review by gottesmm — CHANGES_REQUESTED** [review]
Can you squash these commits as well?

File: cmake/modules/AddSwift.cmake (hunk 2 above)
Comment (gottesmm) [inline cmake/modules/AddSwift.cmake]: @jrose-apple I would like this in a separate commit. I think it is important to explain via the commit message why this change was important.

File: stdlib/CMakeLists.txt (hunk 3 above)
Comment (gottesmm) [inline stdlib/CMakeLists.txt]: "Make a list of "standard library extras" build types that the build should perform.

Or something like that. I think the comment as I wrote it really needs to be clear that you are selecting the specific things to build.
